File: sdk-python/rtmemory/memories.py
"""MemoriesNamespace — async methods for the /v1/memories/ API."""

from __future__ import annotations

from typing import Any

import httpx

from rtmemory.types import (
    Memory,
    MemoryAddRequest,
    MemoryAddResponse,
    MemoryForgetRequest,
    MemoryListResponse,
    MemoryUpdateRequest,
)


class MemoriesNamespace:
    """Namespace for memory CRUD operations."""

    def __init__(self, http: httpx.AsyncClient) -> None:
        self._http = http

    async def add(
        self,
        content: str,
        space_id: str,
        user_id: str | None = None,
        custom_id: str | None = None,
        entity_context: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> MemoryAddResponse:
        """Add a new memory (triggers extraction pipeline)."""
        body = MemoryAddRequest(
            content=content,
            space_id=space_id,
            user_id=user_id,
            custom_id=custom_id,
            entity_context=entity_context,
            metadata=metadata,
        )
        resp = await self._http.post("/v1/memories/", json=body.model_dump(exclude_none=True))
        resp.raise_for_status()
        return MemoryAddResponse.model_validate(resp.json())

    async def list(
        self,
        space_id: str | None = None,
        offset: int = 0,
        limit: int = 20,
    ) -> MemoryListResponse:
        """List memories with pagination and optional filtering."""
        params: dict[str, Any] = {"offset": offset, "limit": limit}
        if space_id is not None:
            params["space_id"] = space_id
        resp = await self._http.get("/v1/memories/", params=params)
        resp.raise_for_status()
        return MemoryListResponse.model_validate(resp.json())

    async def get(self, id: str) -> Memory:
        """Get a single memory by ID (includes version chain)."""
        resp = await self._http.get(f"/v1/memories/{id}")
        resp.raise_for_status()
        return Memory.model_validate(resp.json())

    async def update(
        self,
        id: str,
        content: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Memory:
        """Update a memory's content and/or metadata."""
        body = MemoryUpdateRequest(content=content, metadata=metadata)
        resp = await self._http.patch(f"/v1/memories/{id}", json=body.model_dump(exclude_none=True))
        resp.raise_for_status()
        return Memory.model_validate(resp.json())

    async def forget(
        self,
        memory_id: str | None = None,
        content_match: str | None = None,
        reason: str | None = None,
    ) -> dict[str, Any]:
        """Forget a memory by ID or content match (soft delete)."""
        body = MemoryForgetRequest(memory_id=memory_id, content_match=content_match, reason=reason)
        resp = await self._http.post("/v1/memories/forget", json=body.model_dump(exclude_none=True))
        resp.raise_for_status()
        return resp.json()

File: sdk-js/src/memories.ts
/**
 * MemoryAddClient / MemoryListClient — thin wrappers around the /v1/memories/ API.
 */

import type {
  Memory,
  MemoryAddRequest,
  MemoryAddResponse,
  MemoryForgetRequest,
  MemoryListResponse,
} from "./types";

import { RTMemoryError } from "./client";

/** Helper: build headers from base config. */
function makeHeaders(headers: Record<string, string>): Record<string, string> {
  return { ...headers };
}

// ── MemoryAddClient ───────────────────────────────────────────────

export class MemoryAddClient {
  private baseUrl: string;
  private headers: Record<string, string>;

  constructor(baseUrl: string, headers: Record<string, string>) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.headers = makeHeaders(headers);
  }

  /** Add a new memory (triggers extraction pipeline). */
  async add(req: MemoryAddRequest): Promise<MemoryAddResponse> {
    const res = await fetch(`${this.baseUrl}/v1/memories/`, {
      method: "POST",
      headers: this.headers,
      body: JSON.stringify(req),
    });
    if (!res.ok) throw new RTMemoryError(res.status, await res.text());
    return res.json() as Promise<MemoryAddResponse>;
  }

  /** Forget a memory by ID (soft delete). */
  async forget(memoryId: string, reason?: string): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = {};
    if (reason) body["forget_reason"] = reason;
    const res = await fetch(`${this.baseUrl}/v1/memories/${encodeURIComponent(memoryId)}`, {
      method: "DELETE",
      headers: { ...this.headers, "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    if (!res.ok) throw new RTMemoryError(res.status, await res.text());
    return res.json() as Promise<Record<string, unknown>>;
  }
}

// ── MemoryListClient ──────────────────────────────────────────────

export class MemoryListClient {
  private baseUrl: string;
  private headers: Record<string, string>;

  constructor(baseUrl: string, headers: Record<string, string>) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.headers = makeHeaders(headers);
  }

  /** List memories with pagination and optional space filter. */
  async list(params: {
    spaceId?: string;
    limit?: number;
    offset?: number;
  }): Promise<MemoryListResponse> {
    const url = new URL(`${this.baseUrl}/v1/memories/`);
    if (params.spaceId) url.searchParams.set("space_id", params.spaceId);
    if (params.limit !== undefined) url.searchParams.set("limit", String(params.limit));
    if (params.offset !== undefined) url.searchParams.set("offset", String(params.offset));
    const res = await fetch(url.toString(), {
      method: "GET",
      headers: this.headers,
    });
    if (!res.ok) throw new RTMemoryError(res.status, await res.text());
    return res.json() as Promise<MemoryListResponse>;
  }

  /** Get a single memory by ID. */
  async get(id: string): Promise<Memory> {
    const res = await fetch(`${this.baseUrl}/v1/memories/${id}`, {
      method: "GET",
      headers: this.headers,
    });
    if (!res.ok) throw new RTMemoryError(res.status, await res.text());
    return res.json() as Promise<Memory>;
  }
}
